ambry-store/src/main/java/com/github/ambry/store/Journal.java

```diff
@@ -117,6 +117,40 @@ void addEntry(Offset offset, StoreKey key, Long crc) {
     }
   }
 
+  /**
+   * This method should only be called when replica gets sealed. Remove all entries belongs to the auto closed log segment.
+   * It will delay the addEntry method for sealed replica, so the replication performance will be hold while running this method.
+   * @param activeLogSegmentName journal should remove all entries not belongs to active log segment.
+   */
+  synchronized void cleanUpJournal(LogSegmentName activeLogSegmentName) {
+    Map.Entry<Offset, StoreKey> earliestEntry = journal.firstEntry();
+    while (!inBootstrapMode && earliestEntry != null
+        && earliestEntry.getKey().getName().compareTo(activeLogSegmentName) < 0) {
+      journal.remove(earliestEntry.getKey());
+      recentCrcs.remove(earliestEntry.getValue());
+      currentNumberOfEntries.decrementAndGet();
+      earliestEntry = journal.firstEntry();
+    }
+  }
+
+  /**
+   * This method is used to test the impact of performance between release lock after journal finish clean up and release
+   * lock each time entry gets removed.
+   * @param activeLogSegmentName journal should remove all entries not belongs to active log segment.
+   */
+  void cleanUpJournalWithSynchronizeInside(LogSegmentName activeLogSegmentName) {
+    Map.Entry<Offset, StoreKey> earliestEntry = journal.firstEntry();
+    while (!inBootstrapMode && earliestEntry != null
+        && earliestEntry.getKey().getName().compareTo(activeLogSegmentName) < 0) {
+      synchronized (this) {
+        journal.remove(earliestEntry.getKey());
+        recentCrcs.remove(earliestEntry.getValue());
+        currentNumberOfEntries.decrementAndGet();
+        earliestEntry = journal.firstEntry();
+      }
+    }
+  }
+
```

Comment:
For either `cleanUpJournal` or `cleanUpJournalWithSynchronizeInside`, how can we guarantee the correctness when `getEntriesSince()` is concurrently called by a metadata exchange request?

Reply:
Checked the logic. The getEntriesSince has some protection logic to prevent synchronizing the addEntry method. And I think it also works for cleanUpJournal. if the offset in input is not in journal, it will return null. Which is similar like when Journal size is cross the boundary and the entry mapping to the offset has been removed. So the edge case would be what if the journal size is zero. Inside the getEntriesSince there are logic check if the journal contains the offset at the beginning, so I think it should be good as well.